strategy/json/JSONAPIStrategy.cfc

```diff
@@ -0,0 +1,152 @@
+		ensureDirectory( getOutputDir() );
+		ensureDirectory( getOutputDir() & "/classes" );
+
+		//write the index template
+		var args = {
+			path 		 = getOutputDir() & "/index.json", 
+			template 	 = "#variables.static.TEMPLATE_PATH#/index.cfm", 
+			projectTitle = getProjectTitle()
+		};
+
+		// consider chainable methods, since that's how HTMLAPIStrategy works.
+		var classData = normalize( queryToArray( arguments.qMetadata ) );
+
+		serializeToFile(
+			getOutputDir() & "/index.json",
+			generateClassIndex( classData )
+		);
+
+		classData.each( ( class ) => {
+			serializeToFile(
+				getOutputDir() & "/classes/#class.name#.json",
+				class
+			);
+		});
+
+		return this;
+	}
+
+	/**
+	 * Marshall component metadata into a class index file.
+	 * This will be spit into an `index.json` file at the root of the output directory.
+	 *
+	 * @classData 
+	 */
+	package struct function generateClassIndex( required array classData ){
+		var classMap = arguments.classData.map( ( class ) => {
+			return {
+				"name" : class.name,
+				"path" : "#getOutputDir()#/classes/#class.name#.json"
+			}
+		});
+		return {
+			"packages" : classMap
+		}
+	}
+
+	package array function normalize( required array classData ){
+		var componentMeta = [ "hint", "extends", "functions", "fullname", "properties", "remoteAddress", "type", "hashCode", "name", "path" ];
+		var functionMeta = [ "name", "hint", "description", "access", "closure", "owner", "position", "returnFormat", "returnType", "modifier", "parameters" ];
+		return arguments.classData.map( function( row ) {
+			/**
+			 * Marshall component metadata into attributes array.
+			 */
+			var metaAttributes = row.metadata.filter( ( key ) => {
+				return !arrayContainsNoCase( componentMeta, key );
+			});
+			/**
+			 * Marshall functions to match the designed schema;
+			 * including generating an "attributes" array of attributes not described elsewhere on the function.
+			 */
+			var metaFunctions = row.metadata.functions.map( ( method ) => {
+				return {
+					"returnType" : method.returnType,
+					"returnFormat" : method.returnFormat,
+					"parameters" : method.parameters,
+					"name" : method.name,
+					"hint" : method.keyExists( "hint" ) ? method.hint : "",
+					"description" : method.keyExists( "description" ) ? method.description : "",
+					"access" : method.access,
+					"position" : method.position,
+					"attributes" : method.filter( ( key ) => {
+						return !arrayContainsNoCase( functionMeta, key );
+					})
+				}
+			} );
+			return {
+				"name" : row.name,
+				"package" : row.package,
+				"type" : row.type,
+				"extends" : row.extends,
+				"fullextends" : row.fullextends,
+				"hint" : row.metadata.hint,
+				"attributes" : metaAttributes,
+				"functions" : metaFunctions
+			};
+		});
+	}
+
+
+	/**
+	 * Convert a query object to an array of structs.
+	 *
+	 * @data source query object
+	 */
+	private array function queryToArray( required query data ){
```

Comment:
Why not just do this:

```
query.reduce( (results, row) => { results.append( row ); return results; }, []  )
```

Reply:
:+1: Fantastic suggestion Luis, I went with that. Much cleaner, and tests pass. :smile: